src/qmapshack/gis/CGisListWks.cpp

```diff
@@ -2081,6 +2083,39 @@ void CGisListWks::slotSyncDevWks()
     }
 }
 
+void CGisListWks::slotAddProjectFilter()
+{
+    for(QTreeWidgetItem * item : selectedItems())
+    {
+        IGisProject * project = dynamic_cast<IGisProject*>(item);
+        if(project == nullptr)
+        {
+            continue;
+        }
+        bool hasSearch = false;
+        for(int i = 0; i < project->childCount(); i++)
+        {
+            QTreeWidgetItem* searchItem = project->child(i);
+            //Ideally one could only check at index 0,
+            //but for some reason the search won't stay there
+            if(searchItem->text(0) == tr("Search:"))
+            {
+                hasSearch=true;
+                break;
+            }
+        }
+        if(!hasSearch)
+        {
+            QTreeWidgetItem* searchItem = new QTreeWidgetItem(project, {tr("Search:")});
+            CSearchLineEdit* edit = new CSearchLineEdit(this, project, searchItem);
+            project->insertChild(0, searchItem);
+            project->setSortingFolder(project->getSortingFolder());
```

Comment:
I don't really like this line of code. Any better idea of triggering a sorting to bring the search to the top?

Reply:
Hm, this does not match the way items are added to a project at all. First each item is derived from QTreeWidgetItem (or IGisItem if it is a GIS item). In the case of the filter I would expect something like CProjectFilterItem derived from QTreeWidgetItem. The code to add CSearchLineEdit will be in CProjectFilterItem. To check if there is a filter item you don't have to check for tr("Search:") which might be subject to a change. You dynamic cast to CProjectFilterItem and check for a nullptr. 

And you tell IGisProject to enable or disable the filter item. Let IGisProject handle the necessary code. In IGisProject you have access to all methods. Event to protected sortItems() without doing that nasty hack you don't like,

Reply:
Ah of course :man_facepalming:
Thanks!